homeassistant/components/ring/config_flow.py

```diff
@@ -56,32 +70,87 @@ async def async_step_user(self, user_input=None):
             except Require2FA:
                 self.user_pass = user_input
 
-                return await self.async_step_2fa()
+                return await self.async_step_2fa_user()
 
             except InvalidAuth:
                 errors["base"] = "invalid_auth"
             except Exception:  # pylint: disable=broad-except
                 _LOGGER.exception("Unexpected exception")
                 errors["base"] = "unknown"
 
+        return self._show_user_form(errors)
+
+    async def async_step_2fa_user(self, user_input=None):
+        """Handle 2fa step."""
+        if user_input:
+            return await self.async_step_user({**self.user_pass, **user_input})
+
         return self.async_show_form(
-            step_id="user",
-            data_schema=vol.Schema(
-                {vol.Required("username"): str, vol.Required("password"): str}
-            ),
-            errors=errors,
+            step_id="2fa_user",
+            data_schema=vol.Schema({vol.Required("2fa"): str}),
         )
 
-    async def async_step_2fa(self, user_input=None):
+    async def async_step_2fa_reauth(self, user_input=None):
```

Comment:
It seems like the 2fa_user and 2fa_reauth flows could be combined, where differences are fairly minimal (just the schema?) and be handled by checking for the presence of `self.reauth_entry`.

I would suggest just inlining `_show_reauth_form` and `_show_user_form` -- while the overall method would be more verbose, its easier to understand what is happening if its in a single place without jumping up and down in the file.

Reply:
Both these suggestions are now done.